exporter/prometheusremotewriteexporter/wal.go

```diff
@@ -139,13 +156,136 @@ func (prwe *prweWAL) start(ctx context.Context) error {
 	return nil
 }
 
+func (prwe *prweWAL) run(ctx context.Context) (err error) {
+	if err := prwe.start(ctx); err != nil {
+		return err
+	}
+
+	// Start the process of exporting but wait until the exporting has started.
+	waitUntilStartedCh := make(chan bool)
+	go func() {
+		prwe.continuallyPopWALThenExport(ctx, func() { close(waitUntilStartedCh) })
+	}()
+	<-waitUntilStartedCh
+	return nil
+}
+
+// continuallyPopWALThenExport reads a prompb.WriteRequest proto encoded blob from the WAL, and moves
+// the WAL's front index forward until either the read buffer period expires or the maximum
+// buffer size is exceeded. When either of the two conditions are matched, it then exports
+// the requests to the Remote-Write endpoint, and then truncates the head of the WAL to where
+// it last read from.
+func (prwe *prweWAL) continuallyPopWALThenExport(ctx context.Context, signalStart func()) (err error) {
+	var reqL []*prompb.WriteRequest
+	defer func() {
+		// Keeping it within a closure to ensure that the later
+		// updated value of reqL is always flushed to disk.
+		if errL := prwe.exportSink(ctx, reqL); len(errL) != 0 && err == nil {
+			err = multierror.Append(nil, errL...)
+		}
+	}()
+
+	freshTimer := func() *time.Timer {
```

Comment:
Pardon my ignorance, but why not a ticker?

Reply:
I believe it was done this way so that the timer could be restarted after data is exported.  Using a ticker could result in two consecutive exports, one when the max sample count is reached and another empty, or nearly so, one when the ticker fires shortly thereafter.

Reply:
True, makes sense.